# Penelope-HTG


## Supplementary Resource to:  

Denkert C et al. 2025, Cancer Cell 43, 1â€“16, February 10, 2025 
https://doi.org/10.1016/j.ccell.2025.01.002

Dynamics of molecular heterogeneity in high-risk luminal breast cancer - From intrinsic to adaptive subtyping.

************************************************************

## This resource contains the following data regarding the analyses described in the paper:


### AIMS subtyping based on the HTG gene panel:

[*Subdirectory: AIMS*](https://github.com/tkarn/Penelope-HTG/blob/main/AIMS/):  Calculation of the AIMS subtypes for TCGA samples based on either all genes from the AIMS rules or only on the subset of genes available among the 2549 genes from the HTG EdgeSeq Oncology Biomarker Panel measured on the HTG-Molecular EdgeSeq platform.

************************************************************

### AIMS subtypes in pretherapeutic core biopsies and posttherapeutic surgical samples from paired analysis of 540 tumors before and after neoadjuvant therapy:
[*Subdirectory: Fig2*](https://github.com/tkarn/Penelope-HTG/blob/main/Fig2/):  
Sankey diagrams of paired analysis of 540 tumors before and after therapy and in metastasis showing the transition of AIMS classification during neoadjuvant therapy.

************************************************************

### Differentially expressed genes from paired analysis of 540 tumors before and after neoadjuvant therapy and corresponding sample clustering:
[*Subdirectory: Fig3_FigS2*](https://github.com/tkarn/Penelope-HTG/blob/main/Fig3_FigS2/):  
Code and data to generate Figure 3 and Suppl. Fig.S2 from the paper. Volcano plot of differentially expressed genes from paired analysis of 540 tumors before and after neoadjuvant therapy. Clustered Heatmap containing 1080 tumor samples (540 pre-Tx and 540 post-Tx) as well as 335 differentially expressed genes.

************************************************************

### Relationship of gene clusters and survival:
[*Subdirectory: Fig4*](https://github.com/tkarn/Penelope-HTG/blob/main/Fig4/):  
Analysis comparing the prognostic value of individual genes and their association with gene clusters from unsupervised clustering of Penelope-B samples.

************************************************************

### Gene signature validations and comparisons based on the HTG gene panel

[*Subdirectory:  Fig7_SFig7*](https://github.com/tkarn/Penelope-HTG/blob/main/Fig7_SFig7/):  Analyses and figures comparing Recurrence-Score, GGI, SET-Index in TCGA samples based on all genes and the subset of genes available among the 2549 genes from the HTG EdgeSeq Oncology Biomarker Panel measured on the HTG-Molecular EdgeSeq platform.

************************************************************

### Identification of prognostic groups by additional UMAP analysis:
[*Subdirectory: Fig8*](https://github.com/tkarn/Penelope-HTG/blob/main/Fig8/):  
UMAPs for 1080 tumor samples (540 pre-Tx and 540 paired post-Tx samples) with euclidean metric for the set of 335 differentially regulated genes between pre-Tx and post-TX samples, coloured by AIMS classification and AC subtypes.

************************************************************

### Analysis of prognostic gene sets in pre-Tx biopsies and post-Tx residual tumors:
[*Subdirectory: SFig3*](https://github.com/tkarn/Penelope-HTG/blob/main/SFig3/):  
Normalized Enrichment Scores (NES) of pathways for iDFS prognosis from Gene Set Enrichment Analysis (GSEA) are compared for pre-Tx biopsies and post-Tx residual tumors in a scatter plot.

************************************************************

### Comparison of the genes included in the main Penelope-B gene clusters with the TCGA cohort:
[*Subdirectory: SFig5*](https://github.com/tkarn/Penelope-HTG/blob/main/SFig5/):  
A clustered data table (cdt) format file which can be opened in JavaTreeView. The file contains the median centered RNA-Seq data of the 355 differentially expressed genes for 1215 BRCA samples from TCGA.

************************************************************
